### PR TITLE
feature: add optional argument 'uri_not_comp' for ngx.escape_uri

### PR DESCRIPTION
**Pull request by zhuizhuhaomeng**
I hereby granted the copyright of the changes in this pull request
to the authors of this lua-resty-core project.

related pr:
releated pr:
https://github.com/openresty/lua-resty-core/pull/303
https://github.com/openresty/lua-nginx-module/pull/1709
https://github.com/openresty/meta-lua-nginx-module/pull/77

### CODE REVIEW

File: lib/resty/core/uri.lua [hunk 1]
```diff
@@ -53,22 +53,34 @@ elseif subsystem == "stream" then
...
     end
+
+    if esc_type == nil then
+        esc_type = ngx.ESCAPE_URI_COMPONENT
+    else
+        if type(esc_type) ~= 'number' then
+            error("\"esc_type\" is not number", 3)
+        end
+        if esc_type < ngx.ESCAPE_URI or esc_type > ngx.ESCAPE_MAIL_AUTH then
+            error("\"esc_type\" " .. esc_type .. " out of range", 3)
```
Comment (agentzh) [inline lib/resty/core/uri.lua:72]: Please do not use your internal identifier in the error message. The end user won't have a clue unless they look at the actual source code.

File: lib/resty/core/uri.lua [hunk 2]
```diff
@@ -53,22 +53,34 @@ elseif subsystem == "stream" then
...
         if not s then
             s = ''
         else
             s = tostring(s)
         end
     end
+
+    if esc_type == nil then
+        esc_type = ngx.ESCAPE_URI_COMPONENT
+    else
```
Comment (agentzh) [inline lib/resty/core/uri.lua:67]: Style: needs a blank line before the `else` line.

File: lib/resty/core/uri.lua [hunk 3]
```diff
@@ -53,22 +53,34 @@ elseif subsystem == "stream" then
...
+        if type(esc_type) ~= 'number' then
+            error("\"esc_type\" is not number", 3)
+        end
+        if esc_type < ngx.ESCAPE_URI or esc_type > ngx.ESCAPE_MAIL_AUTH then
```
Comment (agentzh) [inline lib/resty/core/uri.lua:71]: Style: needs a blank line between adjacent `if` statement blocks.

File: lib/resty/core/uri.lua [hunk 4]
```diff
@@ -38,10 +38,10 @@ if subsystem == "http" then
 elseif subsystem == "stream" then
     ffi.cdef[[
     size_t ngx_stream_lua_ffi_uri_escaped_length(const unsigned char *src,
-                                                 size_t len);
+                                                 size_t len, int type);
 
     void ngx_stream_lua_ffi_escape_uri(const unsigned char *src, size_t len,
-                                       unsigned char *dst);
+                                       unsigned char *dst, int type);
```
Comment (agentzh) [inline lib/resty/core/uri.lua:44]: We need PRs for stream lua module otherwise this change will break the stream lua module.

Comment (zhuizhuhaomeng) [reply lib/resty/core/uri.lua:44]: #303
openresty/lua-nginx-module#1709
openresty/meta-lua-nginx-module#77

File: lib/resty/core/uri.lua [hunk 5]
```diff
@@ -53,22 +55,41 @@ elseif subsystem == "stream" then
...
+
+    if esc_type == nil then
+        esc_type = NGX_ESCAPE_URI_COMPONENT
+
+    else
+        if type(esc_type) ~= 'number' then
+            error("\"type\" is not number", 3)
+        end
+
+        if esc_type < NGX_ESCAPE_URI or esc_type > NGX_ESCAPE_MAIL_AUTH then
```
Comment (agentzh) [inline lib/resty/core/uri.lua:76]: If we already check the argument on the lua-resty-core level then we don't need to check it again on the C level in lua-nginx-module.

**Comment by agentzh** [comment]
Merged. Thanks.

**Pull request closed without merging**